### ISSUE TITLE
Profile list saves last chosen profile attributes on failure

### ISSUE THREAD

**Issue by TiPPeX2** [jupyterhub/kubespawner#546]
### Bug description

When failing spawn and choosing a different profile, last profile attributes are saved and not reseted
#### Expected behaviour

a clean profile object is created
#### Actual behaviour

When choosing profile B -> failing spawn -> choosing profile A
some of the attributes from profile B were chosen
### How to reproduce

Using this profile list while having no gpus
```yaml
profileList:
    - display_name: "Super Light"
      description: "2CPU 20GB RAM"
      default: true
      kubespawner_override:
        extra_labels:
          profile: light
          hub.jupyter.org/network-access-hub: 'true'
        cpu_limit: 2
        cpu_guarantee: 2
        mem_limit: 20G
        mem_guarantee: 20G
    - display_name: "Medium GPU"
      description: "1GPU 8CPU 40GB RAM"
      default: false
      kubespawner_override:
        extra_labels:
          profile: beast
          hub.jupyter.org/network-access-hub: 'true'
        cpu_limit: 8
        cpu_guarantee: 8
        mem_limit: 40G
        mem_guarantee: 40G
        extra_resource_limits:
          nvidia.com/gpu: "1"
        extra_resource_guarantees:
          nvidia.com/gpu: "1"
        node_selector:
          company.com/gpu: "true"
        tolerations:
          - effect: NoSchedule
            key: gpu_only
            value: "true"
```
1. Go to spawn 
2. Chose medium GPU profile
3. on failure screen press relaunch
4. chose Super light profile
5. You will still see error due to missing gpu although it's not specified 

### Your personal set up

Using zero-to-jupyterhub
 - OS:
 
 - Version(s):
 
ubunto 18.04
python 3.8.5
- <details><summary>Full environment</summary>

```
alembic==1.4.3
async-generator==1.10
attrs==20.3.0
bcrypt==3.2.0
cachetools==4.1.1
certifi==2020.12.5
certipy==0.1.3
cffi==1.14.4
chardet==3.0.4
cryptography==3.4.4
entrypoints==0.3
escapism==1.0.1
google-auth==1.23.0
idna==2.10
ipython-genutils==0.2.0
Jinja2==2.11.2
jsonschema==3.2.0
jupyter-telemetry==0.1.0
jupyterhub==1.3.0
jupyterhub-firstuseauthenticator==0.14.1
jupyterhub-hmacauthenticator==1.0
jupyterhub-idle-culler==1.0
jupyterhub-kubespawner @ file:///etc/jupyterhub_kubespawner-1.1.2.dev0%2Bcompany.0-py3-none-any.whl
jupyterhub-ldapauthenticator==1.3.2
jupyterhub-ltiauthenticator==1.0.0
jupyterhub-nativeauthenticator==0.0.7
jupyterhub-tmpauthenticator==0.6
kubernetes==12.0.1
ldap3==2.8.1
Mako==1.1.3
MarkupSafe==1.1.1
mwoauth==0.3.7
nullauthenticator==1.0.0
oauthenticator==0.13.0
oauthlib==3.1.0
onetimepass==1.0.1
pamela==1.0.0
prometheus-client==0.9.0
psycopg2-binary==2.8.6
py-spy==0.3.4
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycparser==2.20
pycurl==7.43.0.6
PyJWT==1.7.1
PyMySQL==1.0.2
pyOpenSSL==20.0.0
pyrsistent==0.17.3
python-dateutil==2.8.1
python-editor==1.0.4
python-json-logger==2.0.1
python-slugify==4.0.1
PyYAML==5.3.1
requests==2.25.0
requests-oauthlib==1.3.0
rsa==4.6
ruamel.yaml==0.16.12
ruamel.yaml.clib==0.2.2
six==1.15.0
SQLAlchemy==1.3.23
statsd==3.3.0
text-unidecode==1.3
tornado==6.1
traitlets==5.0.5
urllib3==1.26.2
websocket-client==0.57.0
```
</details>

- <details><summary>Configuration</summary>

```python
Using custom jupyterhub_config.py with LDAP authentication.
```
</details>

- <details><summary>Logs</summary>

```
# paste relevant logs here, if any
```
</details>

**Comment by TiPPeX2** [comment]
Regarding kubespawner package, 
It's a clone obviously from 1.1.2dev with custom change with reflectors.
We wanted to have an ability of single hub with multiple namespace dropdown, we didn't want to have super cluster permissions so we granted manually permission to each namespace we want(pod & event) and changed the reflector singleton to a singleton map basically.

**Comment by manics** [comment]
I've reformatted your description with triple backticks so that it appears as a formatted code block.

Can you confirm this issue still occurs with an unmodified KubeSpawner? Can you also show us your full JupyterHub configuration with secrets redacted, or alternatively reproduce this with a minimal JupyterHub configuration that you can share? Thanks!

**Comment by consideRatio** [comment]
I've reported this also in https://github.com/jupyterhub/jupyterhub/issues/3592 with the original KubeSpawner. I have used hooks though. @TiPPeX2 did you use any hooks like the `pre_spawn_hook`?

**Comment by TiPPeX2** [comment]
@consideRatio 
Yes I'm using the pre_spawn_hook to change spawner.image & spawner.namespace.
Not changing the profiles thought. 

Sounds like this is the same issue so we can close this one, unless you need more info as @manics mentioned.
(Which I might provide later tonight.

I agree it could be a Jupyterhub issue.

**Comment by consideRatio** [comment]
The upstream bug is resolved in JupyterHub, the fix is available in JupyterHub 2.2.0+.
